vivarium/core/engine.py

```diff
@@ -850,22 +860,48 @@ def update(
             self,
             interval: float
     ) -> None:
-        """Run each process for the given interval and update the states.
         """
-
-        time = 0.0
-        emit_time = self.emit_step
+        Run each process for the given interval and update the states.
+        """
         clock_start = clock.time()
+        self.run_for(interval=interval, force_complete=True)
+        self.check_complete()
+        clock_finish = clock.time() - clock_start
+        if self.display_info:
+            self.print_summary(clock_finish)
+
+    def complete(self) -> None:
+        """
+        Force all processes on the front to complete at the current global time
+        """
+        self.run_for(interval=0, force_complete=True)
+        self.check_complete()
 
-        def empty_front(t: float) -> Dict[str, Union[float, dict]]:
-            return {
-                'time': t,
-                'update': {}}
+    def check_complete(self) -> None:
+        """Check that all processes completed
+        """
+        for advance in self.front.values():
+            assert advance['time'] == self.global_time
+            assert len(advance['update']) == 0
+
+    def run_for(
+            self,
+            interval: float,
+            force_complete: bool = False,
+    ) -> None:
+        """
+        Run each process within the given interval and update the states.
 
-        # keep track of which processes have simulated until when
-        front: Dict = {}
+        Args:
+            interval: the amount of time to simulate the composite.
+            force_complete: a bool indicating whether to force processes
+                to complete at the end of the interval.
+        """
+        end_time = self.global_time + interval
+        emit_time = self.global_time + self.emit_step
+        run_complete = False
 
-        while time < interval:
+        while not run_complete:
```

Comment:
Why not just make this condition `self.global_time == end_time`? I think that would be clearer

Reply:
@U8NWXD -- This while loops needs to actually run at `self.global_time == end_time`. So `run_complete` is set to True only after all the processes that need to complete at `end_time` are updated.

Reply:
Then couldn't you make it `self.global_time < end_time`? Or is it possible for `self.global_time` to not be increased?